mewarpx/mewarpx/emission.py

```diff
@@ -0,0 +1,868 @@
+                single number can be passed for it rather than an array.
+        """
+        raise NotImplementedError
+        # data = np.zeros(7)
+
+        # data[0] = warp.top.it*warp.top.dt
+        # data[1] = warp.top.it
+        # data[2] = species.jslist[0]
+        # data[3] = self.getvoltage_e()
+
+        # if E_total is not None:
+        #     if n is not None:
+        #         data[4] = n
+        #     else:
+        #         data[4] = len(E_total)
+
+        #     if w is not None:
+        #         data[5] = species.sq * species.sw * np.sum(w)
+        #     else:
+        #         data[5] = species.sq * species.sw * data[4]
+
+        #     data[6] = np.sum(E_total)
+
+        # self.append_injectedparticles(data)
+
+    def append_injectedparticles(self, data):
+        """Append one or more lines of injected particles data.
+        Arguments:
+            data (np.ndarray): Array of shape (m) or (n, m) where m is the
+                number of fields and n is the number of rows of data to append.
+        """
+        self._injectedparticles_data.append(data)
+
+    def get_injectedparticles(self, clear=False):
+        """Retrieve a copy of injectedparticles data.
+        Arguments:
+            clear (bool): If True, clear the particle data rows entered (field
+                names are still initialized as before). Default False.
+        Returns:
+            injectedparticles_dict (collections.OrderedDict): Keys are the
+                originally passed field strings for lost particles. Values are
+                an (n)-shape numpy array for each field.
+        """
+        raise NotImplementedError
+
+
+class FixedNumberInjector(Injector):
+
+    """Inject n particles every t timesteps."""
+
+    def __init__(self, emitter, species, npart,
+                injectfreq=None, injectoffset=1,
+                weight=0., rseed=None,
+                name=None, unique_particles=True):
+        """Sets up user-specified injection with fixed timestep and weights.
+        Arguments:
+            emitter (:class:`mewarpx.emission.Emitter`): Emitter object that
+                will specify positions and velocities of particles to inject.
+            species (picmi.Species): Premade species to inject particles of.
+            npart (int): Number of particles to inject total
+            injectfreq (int): Number of steps to wait for next injection.
+                Default infinity.
+            injectoffset (int): First timestep to inject. Default 1 (the
+                first possible timestep in WarpX).
+            weight (float): Macroparticle weight to be introduced.
+            rseed (int): If specified, all injection should be repeatable using
+                this rseed. At present each set of injected particles will have
+                the same initial position and velocities as the previous set.
+            name (str or None): Injector name for diagnostics. Constructed from
```

Comment:
```suggestion
            name (str): Injector name for diagnostics. Constructed from
```
I'm not sure if you have to specify or None, since it's not specified for any of the other arguments that can be None.